File: 5-Problem-A-Day/313. Super Ugly Number.java
class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        int[] ugly = new int[n];
        int[] idx = new int[primes.length];
        int[] val = new int[primes.length];
        Arrays.fill(val, 1);

        int next = 1;
        for (int i = 0; i < n; i++) {
            ugly[i] = next;

            next = Integer.MAX_VALUE;
            for (int j = 0; j < primes.length; j++) {
                if (val[j] == ugly[i]) {
                    val[j] = ugly[idx[j]] * primes[j];
                    idx[j] += 1;
                }
                next = Math.min(val[j], next);
            }
        }

        return ugly[n - 1];
    }
}

File: 2021/313. Super Ugly Number.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        int prev = 1;
        vector<int> dp(n, 0);        
        vector<int> indexes(primes.size(), 0);
        dp[0] = 1;
        for (int i = 1; i < n; i++) {
            dp[i] = INT_MAX;
            for (int j = 0; j < primes.size(); j++) {
                dp[i] = min(dp[i], dp[indexes[j]] * primes[j]);                
            }
            for (int j = 0; j < primes.size(); j++) {
                if (dp[i] == dp[indexes[j]] * primes[j]) {
                    indexes[j]++;
                }
            }            
        }
        return dp[n - 1];
    }
};

class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {    
        int pn = primes.size();
        vector<int> pIndexes(pn, 0), nums = { 1 };
        for (int i = 1; i < n; i++) {
            int next = INT_MAX;
            for (int j = 0; j < pn; j++) {
                next = min(next, nums[pIndexes[j]] * primes[j]);                
            }
            for (int j = 0; j < pn; j++) {
                if (next == nums[pIndexes[j]] * primes[j]) {
                    pIndexes[j]++;
                }
            }
            nums.push_back(next);
        }
        return nums[n - 1];
        
    }
};